File: cfr.py
from typing import List, Dict, Type
from functools import reduce

class Game:
    def get_num_players(self) -> int:
        raise NotImplementedError()

class Action:
    pass

class Strategy:
    def __init__(
        self,
        valid_actions
    ):
        pass
    def sample_action(self) -> Action:
        pass


class InfoSet:
    def __init__(
        self,
        valid_actions: List[Action]
    ):
        self.cumulative_regrets: Dict[Action, float] = {action: 0.0 for action in valid_actions}
        self.cumulative_strategy: Dict[Action, float] = {action: 0.0 for action in valid_actions}

        self.current_strategy: Dict[Action, float] = {action: 1/len(valid_actions) for action in valid_actions}

    def update_current_strategy(self):
        normalization = 0
        for action, regret in self.cumulative_regrets:
            self.current_strategy[action] = regret if regret > 0 else 0
            normalization += self.current_strategy[action]

        if normalization > 0:
            for action in self.cumulative_regrets.keys():
                self.current_strategy[action] = self.current_strategy[action] / normalization
        else:
            self.current_strategy = {action: 1 / len(self.current_strategy)}

        
    def get_strategy_prob(self, action: Action) -> float:
        return self.current_strategy[action]

    def get_average_strategy_prob(self, action: Action) -> Strategy:
        pass


class State:
    def __init__(
        self,
        num_agents: int
    ):
        pass

    def get_current_agent(self) -> int:
        pass

    def get_info_set(self) -> InfoSet:
        pass

    def valid_actions(self) -> List[Action]:
        raise NotImplementedError()

    def take_action(self, action: Action) -> State:
        raise NotImplementedError()

    def is_terminal(self) -> bool:
        raise NotImplementedError()

    def get_utility(self, player_id: int) -> float:
        raise NotImplementedError()

    def state_string(self) -> str:
        """
        Should return a unique, minimal string for the current state
        """
        raise NotImplementedError()


class CFR:
    """
    Implements a generic CFR algorithm using the above abstract classes.
    """
    def __init__(
            self,
            game_cls: Type[Game],
            state_cls: Type[State]
        ):
        self.infoSet_map: Dict[str, InfoSet] = {}

        self._state_cls = state_cls
        self.game = game_cls()
        
    def _cfr(self, agent_id, state: State, agent_path_probs: List[float]):
        current_agent = state.get_current_agent()
        
        # Return terminal state payoff
        if state.is_terminal():
            return state.get_utility(agent_id)

        # Get info set
        info_set_key = state.state_string()
        info_set = self.infoSet_map.get(info_set_key)
        if info_set is None:
            info_set = state.get_info_set()

        utility = 0
        action_utils = {}
        for action in state.valid_actions():
            action_prob = info_set.get_strategy_prob(action)
            
            new_path_probs = agent_path_probs.copy()
            new_path_probs[current_agent] = new_path_probs[current_agent] * action_prob

            action_utils[action] = self._cfr(agent_id, state.take_action(action), new_path_probs)
            utility += action_prob * action_utils[action]

        if current_agent == agent_id:
            non_player_prob = 1.0
            for i, prob in enumerate(agent_path_probs):
                if i != agent_id:
                    non_player_prob *= prob

            for action in state.valid_actions():
                info_set.cumulative_regrets[action] += non_player_prob * (action_utils[action] - utility)
                info_set.cumulative_strategy[action] += agent_path_probs[agent_id] * info_set.get_strategy_prob(action)

            info_set.update_current_strategy()

        return utility


    def train(self, iterations: int):
        util = 0
        num_players = self.game.get_num_players()
        for i in range(iterations):
            for player in range(num_players):
                # Get random game init
                random_state = self._state_cls()
                util += self._cfr(player, random_state, [1.0] * num_players)
        print(f'average utility {util/iterations}')

File: cfr/cfr.go
package cfr

import (
	"math"
	"math/rand"
)

type (
	Action     int
	InfoSetKey string
)

// InfoSet ...
type InfoSet struct {
	CumulativeStrategySum map[Action]float64
	CumulativeRegret      map[Action]float64
	CurrentStrategy       map[Action]float64
}

func (info InfoSet) getStateStrategy() map[Action]float64 {
	var norm float64 = 0.0
	stateStrategy := make(map[Action]float64)
	for action, regret := range info.CumulativeRegret {
		if regret > 0 {
			stateStrategy[action] = regret
		} else {
			stateStrategy[action] = 0
		}
		norm += stateStrategy[action]
	}

	for action := range info.CumulativeStrategySum {
		if norm > 0 {
			stateStrategy[action] = stateStrategy[action] / norm
		} else {
			stateStrategy[action] = 1.0 / float64(len(info.CumulativeRegret))
		}
	}
	return stateStrategy
}

func sampleAction(stateStrat map[Action]float64) Action {
	num := rand.Float64()
	sum := 0.0
	for action, likelihood := range stateStrat {
		sum += likelihood
		if sum > num {
			return action
		}
	}
	panic("Unable to sample action for strategy. Maybe strategy is not a valid distribution?")
}

func (info *InfoSet) updateStrategy() {
	normalizingSum := 0.0
	for action, regret := range info.CumulativeRegret {
		info.CurrentStrategy[action] = math.Max(regret, 0)
		normalizingSum += info.CurrentStrategy[action]
	}
	for action := range info.CumulativeRegret {
		if normalizingSum > 0 {
			info.CurrentStrategy[action] /= normalizingSum
		} else {
			info.CurrentStrategy[action] = 1.0 / float64(len(info.CumulativeRegret))
		}
	}
}

func makeInfoSet(ValidActions []Action) InfoSet {
	info := InfoSet{
		CumulativeStrategySum: make(map[Action]float64),
		CumulativeRegret:      make(map[Action]float64),
		CurrentStrategy:       make(map[Action]float64),
	}

	for index := range ValidActions {
		info.CumulativeStrategySum[ValidActions[index]] = 0.0
		info.CumulativeRegret[ValidActions[index]] = 0.0
		info.CurrentStrategy[ValidActions[index]] = 1.0 / float64(len(ValidActions))
	}

	return info
}

type Strategy struct {
	InfoSetMap map[InfoSetKey]InfoSet
}

func NewStrategy() Strategy {
	return Strategy{
		InfoSetMap: make(map[InfoSetKey]InfoSet),
	}
}

type State interface {
	ValidActions() []Action
	// Should return a pointer to the same state object to minimize copying
	TakeAction(Action, bool) State
	// Pass by value so the state is duplicated
	TakeActionCopy(Action) State
	IsTerminal() bool
	GetCurrentAgent() int
	GetUtility(playerID int) float64
	GetInfoSetKey() InfoSetKey
}

func (strat *Strategy) CFR(playerID int, state State, agentPathProbs []float64) float64 {
	return strat.cfr(playerID, state, agentPathProbs, 0)
}
func (strat *Strategy) cfr(playerID int, state State, agentPathProbs []float64, depth int) float64 {
	currentAgent := state.GetCurrentAgent()

	if state.IsTerminal() {
		return state.GetUtility(playerID)
	}

	validActions := state.ValidActions()

	if len(validActions) == 1 {
		// Pass the state by reference since we don't need to run other actions
		return strat.cfr(playerID, state.TakeAction(validActions[0], false), agentPathProbs, depth+1)
	}

	InfoSetKey := state.GetInfoSetKey()
	info, exists := strat.InfoSetMap[InfoSetKey]
	if !exists {
		info = makeInfoSet(validActions)
		strat.InfoSetMap[InfoSetKey] = info
	}

	utility := float64(0.0)
	actionUtility := make(map[Action]float64)
	for _, action := range validActions {
		actionProb := info.CurrentStrategy[action]

		newPathProbs := make([]float64, len(agentPathProbs))
		copy(newPathProbs, agentPathProbs)
		newPathProbs[currentAgent] = newPathProbs[currentAgent] * actionProb

		actionUtility[action] = strat.cfr(playerID, state.TakeActionCopy(action), newPathProbs, depth+1)
		utility += (actionProb * actionUtility[action])
	}

	if currentAgent == playerID {
		// Find the probability of attempting to reach the current state
		nonPlayerPathProb := float64(1.0)
		for index, prob := range agentPathProbs {
			if index != playerID {
				nonPlayerPathProb *= prob
			}
		}

		for _, action := range validActions {
			info.CumulativeRegret[action] += nonPlayerPathProb * (actionUtility[action] - utility)
			info.CumulativeStrategySum[action] += agentPathProbs[playerID] * info.CurrentStrategy[action]
		}

		info.updateStrategy()
	}

	return utility
}
